### DISCUSSION TITLE
operator overloading: where is this going?

### DISCUSSION THREAD

**Discussion by drdbkarron** [dotnet/csharplang#1606] [Q&A] [answered]
I raised this years ago and had to revisit it in my current mathematical c# project.

 As a mathematical programming type, I have had to bite my self--knowing I can't do things that
make c++ wonderful.

What is the status on this enhancement, especially since are you no longer driving or rideing
c# development at Microsoft?

https://stackoverflow.com/questions/1687395/overloading-the-operator-to-add-two-arrays

What's wrong with this C# code? I tried to overload the + operator to add two arrays, but got an error message as follows:

One of the parameters of a binary operator must be the containing type.

class Program
{
  public static void Main(string[] args)
  {
      const int n = 5;

      int[] a = new int[n] { 1, 2, 3, 4, 5 };
      int[] b = new int[n] { 5, 4, 3, 2, 1 };
      int[] c = new int[n];

      // c = Add(a, b);
      c = a + b;

      for (int i = 0; i < c.Length; i++)
      {
        Console.Write("{0} ", c[i]);
      }

      Console.WriteLine();
  }

  public static int[] operator+(int[] x, int[] y)
  // public static int[] Add(int[] x, int[] y)
  {
      int[] z = new int[x.Length];

      for (int i = 0; i < x.Length; i++)
      {
        z[i] = x[i] + y[i];
      }

      return (z);
  }
}

We are considering adding "extension operators" to a hypothetical future version of C#, which would solve your problem. Straw poll: do any of you have AWESOME scenarios for this feature? The more awesome REALISTIC scenarios we can get, the more likely a feature is to be implemented some day. Send 'em my way; you can use the email link on my blog. – Eric Lippert Nov 6 '09 at 15:15

Thanks Eric. Does it mean we will also get "Extension Everything" :) – Joan Venge Nov 6 '09 at 21:26

How do we do this or should I abandon polymorphism until I am polytheistic? – dr d b karron Jun 13 '16 at 5:17

**Comment by svick** [comment, marked as answer]
Extension operators would be part of "extension everything" feature, which is still in an early phase. See https://github.com/dotnet/roslyn/issues/11159 and https://github.com/dotnet/csharplang/issues/192.

**Comment by gafter** [comment]
@drdbkarron You should wrap your array in a custom type. It could be a struct or a class. Define the operators in that new type, and have it return an instance of the new type. That is currently the best way to accomplish what you're trying to do.

Our possible "extension everything" feature may end up being tied together with "type classes" in its design, and therefore we are unlikely to do one without doing the other (for fear we might design ourselves into a corner). C# 8 is possible, but I would guess that C# 9 is more likely.

**Comment by bondsbw** [comment]
Related to your example of element-wise addition (different operator): #431.

Also #823 suggests the `+` operator for concatenating enumerables.

**Comment by msedi** [comment]
I fully agree that we need some "global" mechanism to add additional operators, methods, etc. to existing classes. I was first thinking that something that C++ could do would be the right solution, but I must admit that the "extension everything" approach is the better way.

The only problem I see that for operator overloading, we urgently need the feature also apply mathematical operations on generics. We wrote a large library with a lot of overloads to manage this as this is also the case for System.Math.

**Comment by hypeartist** [comment]
"Mathematical operations on generics" sure would be a really huge thing!

**Comment by drdbkarron** [comment]
Add me in on that 2!

**Comment by lostmsu** [comment]
The feature request generally aligns with our needs to, but is very vague.

**Comment by vbodurov** [comment]
@lostmsu  why is it vague?

```csharp

public static T operator *<T>(T[] a, T[] b) where T :  IMultipliable
{ 
    ....
} 
```
